libraries/AP_Baro/AP_Baro.cpp

```diff
@@ -758,14 +769,31 @@ void AP_Baro::update(void)
     for (uint8_t i=0; i<_num_sensors; i++) {
         if (sensors[i].healthy) {
             // update altitude calculation
+            _atm->consume_updated();
+            _atm->update(i);
+
             float ground_pressure = sensors[i].ground_pressure;
             if (!is_positive(ground_pressure) || isnan(ground_pressure) || isinf(ground_pressure)) {
                 sensors[i].ground_pressure = sensors[i].pressure;
             }
-            float altitude = sensors[i].altitude;
+
+            int8_t alt_mode = _altcalc_mode;
+            float altitude_am = 0;
             float corrected_pressure = sensors[i].pressure + sensors[i].p_correction;
+            switch (alt_mode) {
+            case 0:
+                altitude_am = get_altitude_difference(sensors[i].ground_pressure, corrected_pressure);
+                break;
+            case 1:
+                altitude_am = _atm->get_altitude_amsl();
```

Comment:
You can't provide the absolute altitude here. You're changing the entire frame of reference on the altitudes which all the other libraries don't have insight into, and it would break a *large* amount of code.

Reply:
Thanks for your comments and nop, you are wrong, the old calculation are absolute altitude, because it carry up the ground level every time and introduce an error deviation for altitude calculation, for that reason you can't calculate pressure altitude correctly, plus the fixed temperature added inside the calculations.

Atm has two modes, absolute and relative mode calculation with the QFE and QNH respectively, and it's freeing many unnecessary correction code injected inside the update and optimize the pressure altitude calculation with precision.

get_altitude_amsl() are computed and referenced with the 29.92 inHg presure sea level with actually temperature and not the temeperature and ground level, ground level should be computed trough the QNH and RangeFingers or via ground control or ATC, that is the correct civil aviation APPR operation by the ICAO.

Personally i tested it in flight many times aboard in the plane and METAR data calculation info was provided via COM from Viru Viru ATC Airport for testing QNH and procedures calibrations and the error was 0%, the old calc produces exponential error while ascending altitude.

Try it, carry up your controller board with you aboard into the plane and set the altimeter QNH provided from ATC and fly up to 3500 Feets AMSL and compare it with the airplane instruments with your controller data pressure altitude.

And it was tested with five commercial aircraft pilot instructors from the Fenix Academy for evaluation, the computation follow the TSO 88a FAA indeed.

Reply:
You've misunderstood the comment. You can't change the entire response frame on the call for `AP_Baro::get_altitude()` as this breaks the EKF, and a number of places in the vehicle code. The changes you've made are only sane with a large set of external patches, and outside the intended/support ArduPilot use case.

Reply:
I understand what you mean. EKF isn't affected, the only affected frame are the frontend application when start the system. Sending the command takeoff with AMSL altitude by Mavproxy fly perfectly.